## FERM 533 Assignment II

In [17]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
import numpy as np
import statsmodels.api as sm

import warnings
warnings.filterwarnings("ignore")

#### Let's start with obtaining our data in csv format

In [11]:
firm_data = pd.read_csv("Firm_default.csv")
firm_data.columns = ['Company', 'Default', 'CF_coverage', 'RoA', 'liquidity', 'asset_turnover']

#### I will use the 'CF_Coverage' data in this regression

In [12]:
# A
X_CF_coverage = firm_data[['CF_coverage']]
y = firm_data['Default']
log_reg_CF_coverage = LogisticRegression()
log_reg_CF_coverage.fit(X_CF_coverage, y)
intercept_CF_coverage = log_reg_CF_coverage.intercept_[0]
coef_CF_coverage = log_reg_CF_coverage.coef_[0][0]
regression_equation_CF_coverage = f'logit(P(Default)) = {intercept_CF_coverage:.4f} + ({coef_CF_coverage:.4f}) * CF_coverage'
print(regression_equation_CF_coverage)

logit(P(Default)) = 0.4845 + (-0.0755) * CF_coverage


#### So, what i obtain as equation is; logit(P(Default)) = 0.4845 + (-0.0755) * CF_coverage

In [13]:
# B
X_CF_coverage_sm = sm.add_constant(X_CF_coverage)  
model_CF_coverage_sm = sm.Logit(y, X_CF_coverage_sm)
result_CF_coverage_sm = model_CF_coverage_sm.fit()
print(result_CF_coverage_sm.summary())

Optimization terminated successfully.
         Current function value: 0.475375
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                Default   No. Observations:                   46
Model:                          Logit   Df Residuals:                       44
Method:                           MLE   Df Model:                            1
Date:                Mon, 20 Nov 2023   Pseudo R-squ.:                  0.3104
Time:                        13:29:59   Log-Likelihood:                -21.867
converged:                       True   LL-Null:                       -31.711
Covariance Type:            nonrobust   LLR p-value:                 9.123e-06
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
const           0.4849      0.421      1.151      0.250      -0.341       1.311
CF_coverage    -0.0755    

#### For the coef of CF_coverage, we obtained the negative value -0.0755, which shows an inverse relation between CF_coverage and the probability of a firm defaulting on a loan.
#### The p-value for our CF_coverage is 0.002 < 0.05 significance level, which means that the coefficient is statistically significant.
#### Pseudo R-squared value of 0.3104 prompts that model explains about 31.04% of the variability in the default variable.

In [16]:
# C
X_comprehensive = firm_data[['CF_coverage', 'RoA', 'liquidity', 'asset_turnover']]
X_comprehensive_sm = sm.add_constant(X_comprehensive)
model_comprehensive_sm = sm.Logit(y, X_comprehensive_sm)
result_comprehensive_sm = model_comprehensive_sm.fit()
print(result_comprehensive_sm.summary())

firm_characteristics_new = np.array([40, 3, 300, 100]).reshape(1, -1)
firm_characteristics_new_with_const = np.insert(firm_characteristics_new, 0, 1, axis=1)
log_odds_new = np.dot(result_comprehensive_sm.params, firm_characteristics_new_with_const.T)
probability_of_default_new = 1 / (1 + np.exp(-log_odds_new)).item()
print("Probability of Default for specified firm:", probability_of_default_new)

Optimization terminated successfully.
         Current function value: 0.303869
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:                Default   No. Observations:                   46
Model:                          Logit   Df Residuals:                       41
Method:                           MLE   Df Model:                            4
Date:                Mon, 20 Nov 2023   Pseudo R-squ.:                  0.5592
Time:                        14:01:34   Log-Likelihood:                -13.978
converged:                       True   LL-Null:                       -31.711
Covariance Type:            nonrobust   LLR p-value:                 3.727e-07
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
const              7.6071      3.042      2.501      0.012       1.645      13.569
CF_coverage      

#### We used the values: CF_coverage is 40 , RoA is 3 , liquidity is 300 , asset_turnover is 100.
#### The probability of default is about 0.257%, meaning that company's very unlikely to default, which indicates a low credit risk or a financial stability for these particular parameters. 